src/Microsoft.OData.Core/UriParser/CustomUriFunctions.cs

```diff
@@ -0,0 +1,271 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="CustomUriFunctions.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+namespace Microsoft.OData.Core.UriParser
+{
+    #region NameSpaces
+
+    using System;
+    using System.Collections.Generic;
+    using System.Diagnostics;
+    using System.Linq;
+    using System.Text;
+    using Microsoft.OData.Edm;
+
+    #endregion
+
+    /// <summary>
+    /// Class represents functions signatures of custom uri functions.
+    /// </summary>
+    public class CustomUriFunctions
+    {
+        #region Static Fields
+
+        /// <summary>
+        /// Dictionary of the name of the custom function and all the signatures.
+        /// </summary>
+        private static Dictionary<string, FunctionSignatureWithReturnType[]> CustomFunctions;
+
+        private static object Locker = new object();
+
+        #endregion
+
+        #region Static Ctor
+
+        static CustomUriFunctions()
+        {
+            CustomFunctions = new Dictionary<string, FunctionSignatureWithReturnType[]>();
+        }
+
+        #endregion
+
+        #region Public Methods
+
+        /// <summary>
+        /// Add a custom uri function to extend or override the built-in OData protocol of uri functions.
+        /// In case the function signature already exists as a Built-in function, if requested, the new function signature will override it
+        /// In case the function name already exists as a custom function, the signature will be added as an another overload.
+        /// </summary>
+        /// <param name="customFunctionName">The new custom function name</param>
+        /// <param name="newCustomFunctionSignature">The new custom function signature</param>
+        /// <param name="overrideBuiltInFunction">If 'True', override the existing built-in function in case signature already exists</param>
+        /// <exception cref="ArgumentNullException">Arguments are null, or function signature return type is null</exception>
+        /// <exception cref="ODataException">Throws if built-in function name already exists, and parameter 'overrideBuiltInFunction' is not 'True'</exception>
+        /// <exception cref="ODataException">Throws if built-in function signature overload already exists.</exception>
+        /// <exception cref="ODataException">Throws if custom function signature overload already exists</exception>
+        public static void AddCustomUriFunction(string customFunctionName, FunctionSignatureWithReturnType newCustomFunctionSignature, bool overrideBuiltInFunction = false)
+        {
+            // Parameters validation
+            if (string.IsNullOrEmpty(customFunctionName))

```

Comment:
Use ExceptionUtils.CheckArgumentNotNull(customFuncitonName, "customFunctionName").

same for link 67-70, line 163-166, line 263-266